src/borg/helpers/msgpack.py

```diff
@@ -236,3 +237,21 @@ def int_to_timestamp(ns):
 def timestamp_to_int(ts):
     assert isinstance(ts, Timestamp)
     return ts.to_unix_nano()
+
+
+def assert_normalized_path(path):
+    # `path` should have been normalized earlier. Some features,
+    # like pattern matching relies on a normalized path. As a
+    # precaution we check here again.
+    if to_normalized_path(path) != path:
+        raise AssertionError(f"path {path!r} is not normalized")
+    return path
+
+
+def to_normalized_path(path):
+    assert isinstance(path, str)
+    # Older version of Borg still allowed non-normalized paths
+    # to be stored. So, we normalize them when reading.
+    #
+    # Also checks for '..' elements in `path` for reasons of security.
+    return make_path_safe(path)
```

Comment:
why is this in `helpers.msgpack`?

dealing with old borg stuff is usually done in archiver.transfer_cmd, borg.upgrade, borg.item.

also, please speak of **legacy** rather than "old", so we can later find and remove the not needed any more stuff more easily.

Reply:
> why is this in helpers.msgpack?
>
> dealing with old borg stuff is usually done in archiver.transfer_cmd, borg.upgrade, borg.item.

Yes, so I noticed but this is needed in other code paths too, not just when upgrading. I think we always want to reject '..' and absolute paths and doing that as early as possible seemed safest. ~~Also, there appear to be some places that rely on normalization like #7099 and I think it also messes with pattern matching.~~ (On second though, this only concerns Borg 1. In Borg 2, we'll just normalize all paths during transfer.)

As I understand, Borg 2 only reads v1 repos during `borg transfer`. So, I could move normalization to borg.upgrade but then where should I check for '..' and absolute paths? Still in msgpack?

> also, please speak of legacy rather than "old", so we can later find and remove the not needed any more stuff more easily.

Done

Reply:
> As I understand, Borg 2 only reads v1 repos during borg transfer. So, I could move normalization to borg.upgrade but then where should I check for '..' and absolute paths? Still in msgpack?

Actually, there might be a bit of an issue with absolute paths. During `borg transfer --upgrader=From12To20`, I think we just want to convert absolute paths to relative ones. Since --stdin-name allowed `/a/path`, there might be legitimate archives out there with such paths (and legacy version of Borg may have had other bugs allowing absolute paths). However, we want to reject absolute paths in extract, export-tar, mount and possibly other commands. So, we'd probably want to reject absolute paths early and for all commands (even ones not written yet). Handling this for every command separately will make it more likely that the check is missing somewhere. So, I'd advise against it.

However, this leaves us with one question, where could one put a check for absolute paths that will reject such paths for all commands but `borg transfer --upgrader=From12To20`.

Reply:
I think --stdin-name names (paths) should be just made relative in borg transfer (12to20), like borg usually also makes all other paths relative when creating an archive.

That it did not do that for --stdin-name is rather a bug than a feature.

Also, we should make sure it is a valid pathname or there could be issues with FUSE fs or with extract.

Reply:
Guess the `borg.item.Item.path` attribute is the right place to convert (borg transfer of old stuff) or simply reject (we could do that at same place in borg 2.1 where no transfer of old archive will be supported any more).

Reply:
`borg.item.Item.path` is where I looked first. When I saw that it was defined like this:

```
path = PropDictProperty(str, 'surrogate-escaped str')
```
I figured it would be cleanest to just add a `encode=` and `decode=`.

How would you suggest I handle this instead. I could make `path` an `@property` but then I'd also have to override `__set__` and `__get__` to ensure it's actually called. That seems rather messy though.

I'd appreciate any suggestions as to how to implement this cleanly.

Reply:
keep the encode/decode for now.